In [1]:
import numpy as np

I give the probability at result of softmax will be:
$$softmax(logits) = p$$

for example I give them value of p in case of we have only 6 word in total vocab:
    $$ p =\begin{bmatrix} 0.1 & 0.2 &0.4 &0.1 &0.0 &0.2 \end{bmatrix}$$

and we have the exact result is y :
$$y = \begin{bmatrix} 0 & 0 & 1 & 0 & 0 & 0\end{bmatrix}$$

because y is some constant :
$$changing\_weight\_q\_value\_by =  -\frac {\partial\left(p \right)}{\partial q}$$
$$changing\_weight\_q\_value\_by =  -\frac {\partial\left(softmax(logits) \right)}{\partial logits} \cdot \frac {\partial\left(logits \right)}{\partial q}$$

from definition of error it should be like:
$$Error = {\left(y - p\right)}^2$$

but we use this function for calculate error instead:
$$\mathcal{L} = - \sum_{i = 0}^{V} y_i \log_e \left(p_i\right)$$

call **"cross-entropy loss"**

from the value of y จะได้ว่าค่า $\mathcal{L}$ จะมีค่าเป็น 0 หาก $y = 0$
และ $$\frac{\partial\mathcal{L}}{\partial q} = 0 ; y_i = 0$$ ด้วย

from $$\frac{d}{dx} \log_a(x) = \frac {1} {x \log_e {a}} = \frac {1} {x \ln\left(a\right)}$$

if we want to decrease error by change some value that will be combine into p 
we will write it to :
$$ changing\_weight\_q\_value\_by = \frac {\partial\mathcal{L}}{\partial q}$$
$$ \frac {\partial\mathcal{L}}{\partial q} = \frac {\partial}{\partial q} \left(-\sum_{i = 0}^V \left( y_i \ln \left(p_i\right) \right)\right)$$


from the y array we found that the equation have the value only at $y_i = 1$ then we got 
$$ \frac {\partial\mathcal{L}}{\partial q} = -\sum_{i=0}^V\frac {\partial\left( y_i \ln(p_i)\right)}{\partial p_i} \cdot \frac{\partial p_i} {\partial q}$$

$$ \frac {\partial\mathcal{L}}{\partial q} = -\frac {y_i}{p_i \ln\left(e\right)} \cdot \frac{\partial p_i} {\partial q}$$

## Softmax derivation
---

from softmax equation:
$$p_i = softmax(z_i) = \frac{e^{z_i}}{\sum_{k = 0}^{V - 1} \left(e^{z_k}\right)}$$

**การหาอนุพันธ์ของ Softmax จะพิเศษกว่าขั้นตอนอื่นเนื่องจาก $softmax\left(z_i\right)$ ไม่ได้ขึ้นอยู่กับแค่ $z_i$ แต่จะขึ้นอยู่กับค่าของ logits ทุกตัว
ดังนั้น**

if $$logits = \begin{bmatrix}z_0&z_1&...&z_j&...&z_{V - 1}\end{bmatrix}$$

$$\frac{\partial p_i} {\partial z_m} = ?$$
if $m = \begin{matrix} 0, 1,..., V-1 \end{matrix}$

We start with :

$$\frac{d}{dz_m} \frac{U}{S} = \frac{U^\prime S - US^\prime}{S^2}$$



$$U = e^{z_i}$$
and 
$$S = \sum_{k=0}^S \left(e^{z_k}\right)$$

$$\frac {\partial U} {\partial z_m} = 
\begin{cases} 
    e^{z_i} & \text{if $i = m$}\\ 
    0 & \text {if $i \neq m$} 
\end{cases}$$
**Or**
$$\frac {\partial U} {\partial z_m} = e^{z_i}\delta_{i,m}$$

then in case of $i = m$ we got : 

$$\frac{\partial p_i}{\partial z_m} = \frac{e^{z_i} S - e^{2\cdot z_i}}{S^2}$$
$$\frac{\partial p_i}{\partial z_m} = \frac{e^{z_i}}{S} \cdot \frac{S - e^{z_i}}{S}$$

finally got
$$\frac{\partial p_i}{\partial z_m} = p_i \cdot \left(1 - p_i\right)$$

if $i \neq m$ :
$$
    \frac{\partial p_i}{\partial z_m} = \frac{0 \cdot S - e^{z_i}e^{z_m}}{V^2}
$$

$$
    \frac{\partial p_i}{\partial z_m} = -\frac{e^{z_i}e^{z_m}}{S^2}
$$


**Or**
$$\frac{\partial p_i}{\partial z_m} = p_i \cdot \left(\delta_{i,m} - p_i\right)$$


---

## Mixing up

We have
$$\frac{\partial p_i}{\partial z_m} = p_i \cdot \left(\delta_{i, m} - p_j\right)$$
และ loss function (cross-entropy) ต่อ 1 ตำแหน่ง:

$$\mathcal{L} = - \sum_{i = 0}^{V} y_i \log_e \left(p_i\right)$$

โดย y มักจะเป็น one-hot (แต่ derivation นี้ใช้ได้กับ distribution ด้วย)

#### 1.) เริ่ม chain rule

$$\frac {\partial \mathcal {L}} {\partial z_m} =
\sum_{j=0}^{V - 1} \frac{\partial \mathcal{L}}{\partial p_i} \cdot
\frac {\partial p_i}{\partial z_m}$$

หา $\frac {\partial \mathcal{L}}{\partial p_i}$:

$$\frac {\partial \mathcal{L}}{\partial p_i} = -\frac{y_i}{p_i}$$

แทนกลับเข้าไป :
$$\frac {\partial \mathcal {L}} {\partial z_m} =
\sum_{i=0}^{V - 1} \left(\left(-\frac{y_i}{p_i} \right)\cdot
p_i \cdot \left(\delta_{i, m} - p_i\right)\right)

ตัด $p_i$ ทิ้ง :
$$\frac {\partial \mathcal {L}} {\partial z_m} =
\sum_{j=0}^{V - 1} \left(-{y_i}  \cdot \left(\delta_{i, m} - p_i\right)\right)

#### 2.) แยกผลรวมออกเป็น 2 ก้อน (นี่คือจุดที่ "เคสอื่น ๆ" จะถูกกินรวม)

$$-\sum_i y_i\left(\delta_{i, m} - p_m\right) = - \sum_i y_i \delta_{i, m} + \sum_i y_i p_m$$

สังเกตว่า $p_m$ ไม่ขึ้นกับ $i$ (สำหรับการ sum ตาม $i$) จึงสามารถดึงออกมาจากผลรวมได้:

$$ \frac {\partial \mathcal{L}} {\partial z_m} = -\sum_i y_i \delta_{i, m} + p_m \sum_i y_i$$

#### 3.) ประเมินผลรวมของ $\sum_i y_i \delta_{i,m}$ และ $\sum_i y_i$

* $\sum_i y_i \delta_{i, m}$ : $\delta_{i, m}$ ทำหน้าที่ "เลือก" เฉพาะตำแหน่ง $i = m$ ดังนั้น : 
$$ \sum_i y_i \delta_{i, m} = y_m$$

* $\sum_i y_i$ : ไม่ว่า $y$ เป็น one-hot หรือจะเป็น probability distribution จะมีค่าผลรวมเท่ากับ 1 เสมอ ดังนั้น :
$$ \sum_i y_i = 1$$


#### 4.) สรุปออกมาเป็น $p-y$

แทนค่า : 
$$\frac{\partial \mathcal {L}} {\partial z_m} = -y_m + p_m \cdot 1 = p_m - y_m$$

ดังนั้นในรูป vector : 
$$\frac{\partial \mathcal {L}} {\partial \vec{z}} = \vec{p}- \vec{y}$$

---

## Learn from dummy

สมมุติในการเทรน LLM เราจะมีข้อความหนึ่งชุด แล้วตัดเป็นชิ้นยาว T = 8 tokens:
* ถ้าใน 1 step เราใช้ข้อความแค่ชิ้นเดียว -> Batch size $B = 1$
    * input shape: (1, 8)
    * model จะทำนาย next-token 8 ตำแหน่ง
    * จำนวน token ที่ฝึกต่อ step = $B \cdot T = 8$ 

ตัวอย่าง batch จริง ๆ (หลายชิ้นพร้อมกัน)

สมมติเราเตรียมชิ้นข้อความ (แต่ละชิ้นยาว 8 tokens) ได้ 4 ชิ้น : 
* Step หนึ่งใช้ 4 ชิ้นพร้อมกัน -> Batch size $B = 4$
    * input shape: (4, 8)
    * จำนวน token ที่ฝึกต่อ step = $4 \cdot 8 = 32$

**ผลที่ได้คือ** : gradient ที่ได้ต่อ step จะมีความนิ่งขึ้น เพราะมาจากตัวอย่างที่มากขึ้น


* ตอนที่เราทำการเทรนทีละ Sequence หรือ ฺ$B = 1$ ใช้สมการเดิมได้ไม่มีปัญหา

* แต่ถ้าเอา 32 sequence มารวมกันเป็น batch -> $B = 32$

ทำให้ shape ที่ได้เป็น : 

* logits $Z: (B, T, V)$
* probs $P: (B, T, V)$
* targets $Y: (B, T, V)$ (one-hot) หรือ (B, T)(index)

---

ทำให้ตอนคิด Loss function จะต้องคิดเป็นค่าเฉลี่ย (mean) (เป็นที่นิยมใช้) แล้วได้สมการเป็น

$$\mathcal {L} = \frac {1} {B \cdot T} \sum_{b=1}^B \sum_{t=1}^T \left(-ln \left(p_{b, t, y_{b,t}}\right)\right)$$

แล้วทำให้ gradient มี factor $\frac {1} {B \cdot T}$ ติดมาด้วย : 

$$\frac{\partial \mathcal{L}} {\partial \vec{Z}} = \frac{\vec{P} - \vec{Y}}{B \cdot T} $$

In [2]:
def softmax(x, axis = -1):

    # หาค่ามากสุดตามแนวแกนที่เลือกมาแล้วหักลบกับแต่ละค่าในแมทริกส์
    # axis = -1 : [1, 2, 3], [4, 5, 6] -> (2, 3) shape 
    # ได้คำตอบเป็น [[3], [6]] ที่เป็นค่ามากสุดแต่ละค่าตามแนวแกน X
    x = x - np.max(x, axis = axis, keepdims=True)
    e = np.exp(x)
    return e / np.sum(e, axis = axis, keepdims = True)

### 1) for example Vocab มี 6 ตัว


In [3]:

vocab = ["<PAD>", "I", "You", "love", "a", "cat"]

มีทั้งหมด 2 ประโยค (Batch size $B = 2$)

In [4]:
sentenceA = "I love a cat"
sentenceB = "You love a cat"

เราจะตั้ง sequence length ที่ใช้เทรนเป็น $T = 3$ (เอาแค่ 3 ตัวแรกเป็น input)

### 2 สร้าง X (input) และ Y (output) ด้วยการเลื่อนไปทางซ้าย 1 ระดับ 

In [5]:
X_a = [1, 3, 4]
Y_a = [3, 4, 5]

X_b = [2, 3, 4]
Y_b = [3, 4, 5]

### 3) เขียนเป็น Matrix (B, T)

In [6]:
X = np.array([X_a, X_b]);
Y = np.array([Y_a, Y_b]);

### 4) Model ให้ logits ออกมาเป็น (B, T, V)

$Z$ หรือ logits หมายถึงคะแนนดิบก่อน softmax
$$ Z \in \mathbb{R}^{2 \times 3 \times 6}$$
ในกรณีนี้
* มี ฺBatch size = 2 
* มี token size = 3 
* มี Vocab size (V) = 6


In [7]:
Z_0 = np.array([
    [-1, .2, -.3, 2.5, .1, -.7],
    [-.4, -.2, -.5, .1, 2.2, 0],
    [-.8, -.3, -.2, 0, .4, 2.6]
    ])



In [8]:
Z_1 = np.array([
    [-1.1, -.1, 0, 2.3, .2, -.6], 
    [-.5, -.3, -.4, .2, 2.4, -.1],
    [-.9, -.4, -.1, .1, .3, 2.5]
    ])

In [9]:
Z = np.array([Z_0, Z_1])
Z.shape

(2, 3, 6)

### 5) softmax ให้ได้รูปแบบ $P (B, T, V)$

In [10]:
P = softmax(Z)
print(P)

[[[0.02282931 0.07579597 0.04597258 0.75600279 0.06858303 0.03081634]
  [0.05068289 0.06190422 0.04585978 0.08356196 0.68238116 0.07560999]
  [0.02501213 0.04123803 0.04557508 0.05566552 0.0830432  0.74946603]]

 [[0.02380694 0.06471396 0.07151999 0.7133534  0.08735471 0.039251  ]
  [0.03998993 0.04884381 0.04419571 0.08052983 0.72678277 0.05965796]
  [0.02446378 0.04033395 0.05444514 0.06649944 0.0812226  0.73303509]]]


In [11]:
P.shape

(2, 3, 6)

**Structure of P และทำไมถึงใช้วิธีการหยิบค่ามาคำนวณได้เลย**

current output of P consist of 2 batch (index = 0 of shape)
ความน่าจะเป็นของโทเคนที่จะออกถัดไปของแต่ละ x ใน index = 1 of shape 
* ยกตัวอย่างของ batch ที่ 1 (P[0]) เราจะได้ว่าผลการเดาโทเคนตัวต่อไปของคำว่า I คือ (love, มีค่าความน่าจะเป็น P[0, 0, 3] = 0.75600279) 
    - แต่จาก one hot Y vector คำตอบที่ถูกต้องจะเป็น $Y = \begin{bmatrix} 0 & 0 & 0 & 1 & 0 & 0 \end{bmatrix}$
    - จากสมการเต็ม
    $$ \mathcal{L} = \frac{1} {B \cdot T} \sum_{b=0}^{B-1} \sum_{t=0}^{T-1} y_t \ln \left(p_t\right)$$
    - แต่ที่ $Y_{t \neq ตำแหน่งคำตอบที่ถูกต้อง} = 0$ จึงทำให้สมการลดรูปเหลือ
    $$ \mathcal{L} = \frac{1} {B \cdot T} \sum_{b, t} - \ln \left(P[b, t, Y[b, T]]\right)$$

### 6) คิด loss ด้วย Y (target)

เพราะ Y เก็บเป็น id ไม่ใช่ one-hot ทำให้เราคิด cross-entropy แบบหยิบเฉพาะคำ

$$ \mathcal{L} = \frac{1} {B \cdot T} \sum_{b, t} -ln \left(P[b, t, Y[b, t]]\right)$$

ตัวอย่างการหยิบค่ามาคิด loss

* สำหรับ A ที่ $t=0$ : target คือ $Y[0, 0] = 3$ (love) -> ใช้ค่า $P[0, 0, 3]$
* สำหรับ A ที่ $t=1$ : target คือ $Y[0, 1] = 4$ (a) -> ใช้ค่า $P[0, 1, 4]$ 

In [12]:
print(Y[0, 0])
print(P[0, 0, 3])

3
0.7560027866558934


In [13]:
B = P.shape[0] #total batch
T = P.shape[1] #total token in each step

totalLoss = 0
for b in range(B):
    for t in range(T):
        totalLoss += - np.log(P[b, t, Y[b, t]])
Loss = totalLoss / (B * T)

print("Loss ของรอบนี้คือ ", Loss)

Loss ของรอบนี้คือ  0.31962317909320725


In [14]:
B = 5
T = 6
V = 20

p = np.array([
    [.1, .2, .4, .1, 0, .1], 
    [.1, .1, .2, 0, 5, .1]]);
y = np.array([
    [0, 0, 1, 0, 0, 0], 
    [0, 0, 0, 0, 1, 0]
    ]);

$$\text{softmax}\left(\text{logits}\right) =  \frac{e^{logits_i-logits_{max}}}{\sum_{i = 0}^{V-1} e^{logits_i-logits_{max}}}$$

When $V$ is vocab size

### 7) แล้ว Gradient of logits จะเขียนได้ดังนี้

$$ \frac {\partial \mathcal {L}} {\partial Z} = \frac {P - Y_{onehot}} {B \cdot T} $$

วิธีด้านล่างนี้ผิด ==>

In [15]:
#find gradient of logits
B = P.shape[0]
T = P.shape[1]


# วิธีนี้ผิด
totalGradientOfLogits = 0

for b in range(B):
    for t in range(T):
        totalGradientOfLogits += P[b, t, Y[b, t]] - Y[b,t]

GradientOfLogits = totalGradientOfLogits / (B * T)
print ("Gradient of logits ในการคำนวณรอบนี้คือ ", GradientOfLogits)

Gradient of logits ในการคำนวณรอบนี้คือ  -3.2731631266786043


#### วิธีที่ถูกต้อง ==>

In [16]:
# แบบใช้ one hot vector Y

B, T = Y.shape
V = P.shape[2]

Y_onehot = np.zeros((B, T, V), dtype=np.float64)
Y_onehot[np.arange(B)[:, None], np.arange(T)[None, :], Y] = 1.0

print("ได้ Onehot of Y เป็น")
print(Y_onehot)

ได้ Onehot of Y เป็น
[[[0. 0. 0. 1. 0. 0.]
  [0. 0. 0. 0. 1. 0.]
  [0. 0. 0. 0. 0. 1.]]

 [[0. 0. 0. 1. 0. 0.]
  [0. 0. 0. 0. 1. 0.]
  [0. 0. 0. 0. 0. 1.]]]


In [17]:
GradientOfLogits = (P - Y_onehot) / (B * T)

print("แล้วได้ Gradient of logits เป็น ")
print(GradientOfLogits)

แล้วได้ Gradient of logits เป็น 
[[[ 0.00380488  0.01263266  0.0076621  -0.0406662   0.0114305
    0.00513606]
  [ 0.00844715  0.01031737  0.0076433   0.01392699 -0.05293647
    0.01260166]
  [ 0.00416869  0.00687301  0.00759585  0.00927759  0.01384053
   -0.04175566]]

 [[ 0.00396782  0.01078566  0.01192    -0.04777443  0.01455912
    0.00654183]
  [ 0.00666499  0.00814063  0.00736595  0.01342164 -0.04553621
    0.00994299]
  [ 0.0040773   0.00672233  0.00907419  0.01108324  0.0135371
   -0.04449415]]]


จะเห็นได้ว่า gradient ของ lost function และ softmax (gradient ของ logits) $\frac {\partial \mathcal {L} } {\partial P} \cdot \frac {\partial P} {\partial Z}$ รวมกันมีมิติ เท่ากับ logits

$$\frac {\partial \mathcal {L}} {\partial Z} \in \mathbb {R} ^ {B \times T \times V}$$

Where 
* $B$ is batch size
* $T$ is token size
* $V$ is vocab size

In [18]:
# index target method

B, T, V = P.shape

dZ = P.copy()
dZ /= (B * T)

# subtract 1/(B * T) at the true class indices

dZ[np.arange(B)[:, None], np.arange(T)[None, :], Y] -= 1.0 / (B * T)

print (dZ.shape)
print(dZ)

(2, 3, 6)
[[[ 0.00380488  0.01263266  0.0076621  -0.0406662   0.0114305
    0.00513606]
  [ 0.00844715  0.01031737  0.0076433   0.01392699 -0.05293647
    0.01260166]
  [ 0.00416869  0.00687301  0.00759585  0.00927759  0.01384053
   -0.04175566]]

 [[ 0.00396782  0.01078566  0.01192    -0.04777443  0.01455912
    0.00654183]
  [ 0.00666499  0.00814063  0.00736595  0.01342164 -0.04553621
    0.00994299]
  [ 0.0040773   0.00672233  0.00907419  0.01108324  0.0135371
   -0.04449415]]]


## Padding and loss mark

Loss mark คือ ตัวกรองว่า ตำแหน่งไหน ใน Sequence ที่เราจะนำไปคิด Loss และตำแหน่งไหน ไม่คิด  

**โดยจะใช้ตอนคำนวณ Loss เท่านั้นไม่คิดใน attention**

**Loss mark ใช้ทำอะไรบ้าง**

1. กัน PAD ไม่ให้ถูกเอาไปคิด loss

เวลา batch ประโยคยาวไม่เท่ากัน เราต้อง pad ให้ยาวเท่ากันเช่นการใส่ <PAD> = id 0 ตำแหน่งที่ PAD ไม้ใช่ข้อมูลจริง -> *ไม่ควรคิด loss*

2. Instruction tuning / Chat format

เรามัก ไม่อยากให้โมเดลเรียนรู้การทวน Prompt เราอยากให้เรียนเฉพาะคำตอบ (assistanct)
เลย mark ตำแหน่ง prompt = 0 และคำตอบ = 1

3. Ignore special tokens

เช่น <BOS> <SEP> หรือบางคำที่ไม่อยากให้ train


### Shape of loss mark 

loss mark มักจะมี shape เดียวกับ target tokens:
* $Y \in \mathbb {R} ^ {B \times T}$
* $loss\_mark \in \left\{1, 0\right\} ^ {B \times T}$ และมีค่าข้างในแค่ 0 และ 1

### สูตรคำนวณ loss แบบมี mask 

$$\mathcal {L} = \frac {\sum_{b,t} m_{b, t} \cdot \left(-log \left( P [b, t, Y[b, t]]\right)\right)} {\sum_{b, t} m_{b, t}}$$

### 1) Demo data
---
* token ที่เป็น system / user / prompt &rarr; mask = 0
* token ที่เป็น assistant response &rarr; mask = 1

ในขั้นตอนของการ tokenization ถ้ารู้ช่วง index ของ assistant เช่นช่วง $\left[ start : end \right)$ สามารถใช้คำสั่ง

In [19]:
#focus ที่ mask y

start = 0;
end = 2
loss_mask = np.zeros((B, T), dtype=np.float64)
loss_mask[b, start: end] = 1

print(loss_mask)

[[0. 0. 0.]
 [1. 1. 0.]]


In [20]:
X_mask1 = [3, 4, 5] #love a cat
Y_mask1 = [4, 5, 0] #a cat <PAD>

X_mask2 = [4, 5, 0] #a cat <PAD>
Y_mask2 = [5, 0, 0] #cat <PAD> <PAD>

pad_id = 0;

In [21]:
Y = np.array([Y_mask1, Y_mask2]);
print(Y)

[[4 5 0]
 [5 0 0]]


loss_mask (1 เฉพาะตำแหน่งที่ไม่ใช่ PAD)

> * m คือ loss mask matrix
> * M คือ จำนวน token ทั้งหมดที่จะถูกนำไปเทรน

กติกา: ตำแหน่ง Y == pad_id -> mask = 0 (ไม่คิด loss)
ตำแหน่งที่ Y != pad_id -> mask = 1 (คิด loss)

จำนวนตำแหน่งที่คิด loss ทั้งหมด
$$M = \sum loss\_mask = 3$$


In [22]:

# ถ้าค่าใน Y ไม่เท่ากับ pad_id (0) จะได้เป็นค่า 1 นอกนั้น 0
m = (Y != pad_id).astype(np.float64) # (B, T)
M = m.sum()

print(m)

[[1. 1. 0.]
 [1. 0. 0.]]


สมมุติ logits $ Z \in \mathbb {R} ^ {B \times T \times V}$ 

**ปรกติเลขที่ได้จะไม่ใช่ 0 ล้วนแต่เพื่อให้ง่ายต่อการมองเห็นภาพ โดยปรกติแล้วตำแหน่งของเลขที่มากที่สุดจะหมายถึง ความน่าจะเป็นที่ token ถัดไปจะออก**

In [23]:
Z = np.zeros((B, T, V), dtype=np.float64)

Z[0, 0, 4] = 1.0 # (b0, t0) target 4
Z[0, 1, 5] = 2.0 # (b0, t1) target 5
Z[1, 0, 5] = 0.5 # (b1, t0) target 5

print(Z)

[[[0.  0.  0.  0.  1.  0. ]
  [0.  0.  0.  0.  0.  2. ]
  [0.  0.  0.  0.  0.  0. ]]

 [[0.  0.  0.  0.  0.  0.5]
  [0.  0.  0.  0.  0.  0. ]
  [0.  0.  0.  0.  0.  0. ]]]


นำค่า logits ที่ได้ไปหาค่าความน่าจะเป็น $P$ โดยใช้ softmax function &rarr;

In [24]:
P = softmax(Z)
print(P)

[[[0.12956251 0.12956251 0.12956251 0.12956251 0.35218743 0.12956251]
  [0.0807164  0.0807164  0.0807164  0.0807164  0.0807164  0.596418  ]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]]

 [[0.15040486 0.15040486 0.15040486 0.15040486 0.15040486 0.24797569]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]]]


one-hot matrix Y &rarr; Y_onehot

In [25]:
Y_onehot = np.zeros((B, T, V), dtype=np.float64)
Y_onehot[np.arange(B)[:, None], np.arange(T)[None, :], Y] = 1.0

print(Y_onehot)

[[[0. 0. 0. 0. 1. 0.]
  [0. 0. 0. 0. 0. 1.]
  [1. 0. 0. 0. 0. 0.]]

 [[0. 0. 0. 0. 0. 1.]
  [1. 0. 0. 0. 0. 0.]
  [1. 0. 0. 0. 0. 0.]]]


In [26]:
print(vocab)

['<PAD>', 'I', 'You', 'love', 'a', 'cat']


[[[0. 0. 0. 0. 1. 0.] >> คำต่อไปที่จะออกคือ "a"  
  [0. 0. 0. 0. 0. 1.] >> คำต่อไปที่จะออกคือ "cat"  
  [1. 0. 0. 0. 0. 0.]] >> คำต่อไปที่จะออกคือ "PAD"  

 [[0. 0. 0. 0. 0. 1.]  
  [1. 0. 0. 0. 0. 0.]  
  [1. 0. 0. 0. 0. 0.]]]  

In [27]:

#สร้าง array ใหม่โดยใช้วิธีการหยิบค่าจาก p_correct โดยที่

#p_correct[0, 0] หมายถึงเข้าไปหยิบค่ามาจาก P[0, 0, Y[0, 0]]
#p_correct[1, 2] หมายถึงเข้าไปหยิบค่ามาจาก P[1, 2, Y[1, 2]]
p_correct = P[np.arange(B)[:, None], np.arange(T)[None, :], Y] # (B, T)
print("p_correct ->\n", p_correct);

nll = -np.log(p_correct + 1e-12) # เลข 1e-12 ใส่มาเพื่อกัน log (0)
loss = (nll * m).sum() / max(M, 1)
print("loss : ", loss)

p_correct ->
 [[0.35218743 0.596418   0.16666667]
 [0.24797569 0.16666667 0.16666667]]
loss :  0.98494327816074


#### วิธีการคำนวน A : $\frac{\partial \mathcal {L} } {\partial Z} = \frac {\left(P - Y_{one-hot}\right)\ *\ m[:, :, None] } {M}$

In [28]:
dZ = (P - Y_onehot) * m[:, :, None]
dZ = dZ / max(M, 1)
print (dZ)

[[[ 0.0431875   0.0431875   0.0431875   0.0431875  -0.21593752
    0.0431875 ]
  [ 0.02690547  0.02690547  0.02690547  0.02690547  0.02690547
   -0.13452733]
  [-0.          0.          0.          0.          0.
    0.        ]]

 [[ 0.05013495  0.05013495  0.05013495  0.05013495  0.05013495
   -0.25067477]
  [-0.          0.          0.          0.          0.
    0.        ]
  [-0.          0.          0.          0.          0.
    0.        ]]]


In [29]:
print("loss_mark m = \n", m) 
print("M = ", M)
print("masked_loss = ", loss)

print("\ndZ[0, 0, :] =", dZ[0, 0, :])
print("dZ[0, 1, :] =", dZ[0,1,:])
print("dZ[0, 2, :] =", dZ[0,2,:])  # PAD target -> mask=0 -> should be all zeros

loss_mark m = 
 [[1. 1. 0.]
 [1. 0. 0.]]
M =  3.0
masked_loss =  0.98494327816074

dZ[0, 0, :] = [ 0.0431875   0.0431875   0.0431875   0.0431875  -0.21593752  0.0431875 ]
dZ[0, 1, :] = [ 0.02690547  0.02690547  0.02690547  0.02690547  0.02690547 -0.13452733]
dZ[0, 2, :] = [-0.  0.  0.  0.  0.  0.]


จะเห็นได้ว่าตำแหน่งที่ถูก mask ไว้ (Y = PAD) &rarr; $m = 0$  &rarr; $dZ = 0$  

ทำให้ทั้ง Vocab &rarr; weight upstream ไม่โดน Update จากตำแหน่งนั้นเลย

#### วิธีการคำนวณ B : index trick

In [30]:
Y

array([[4, 5, 0],
       [5, 0, 0]])

In [31]:
dZ = P.copy()
B, T, V = P.shape
M = m.sum()


print("P = \n", P)


# subtract 1 at the true class indices
dZ[np.arange(B)[:, None], np.arange(T)[None, :], Y ] -= 1.0

# apply mask and normalize
dZ *= m[:, :, None]
dZ /= max(M, 1) # avoid division by zero
print("\n\ndZ => \n", dZ)

P = 
 [[[0.12956251 0.12956251 0.12956251 0.12956251 0.35218743 0.12956251]
  [0.0807164  0.0807164  0.0807164  0.0807164  0.0807164  0.596418  ]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]]

 [[0.15040486 0.15040486 0.15040486 0.15040486 0.15040486 0.24797569]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]
  [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]]]


dZ => 
 [[[ 0.0431875   0.0431875   0.0431875   0.0431875  -0.21593752
    0.0431875 ]
  [ 0.02690547  0.02690547  0.02690547  0.02690547  0.02690547
   -0.13452733]
  [-0.          0.          0.          0.          0.
    0.        ]]

 [[ 0.05013495  0.05013495  0.05013495  0.05013495  0.05013495
   -0.25067477]
  [-0.          0.          0.          0.          0.
    0.        ]
  [-0.          0.          0.          0.          0.
    0.        ]]]


### ทำไมตอนใส่ loss mask เราใช้แค่ M ในการ normalize ไม่ใช่ B * T หรือ M * B * T

> เป้าหมายในการ normalize คือให้ loss / gradient เป็น average ต่อ 1 token ที่ถูก train ซึ่งจำนวน token ที่ถูก train คือ $M = \sum m$

**กันสับสน ตำแหน่งที่ถูก train จะมีค่าเท่ากับ 1 ใน loss mask matrix**

In [32]:
loss_mask = np.array([[1, 1, 0], [1, 0, 0]])
print(loss_mask)

[[1 1 0]
 [1 0 0]]


In [33]:
def lm_head_forward(H, W, b):
    #H: (T, d), W: (d, V), b: (V,)
    Z = H @ W + b # (T, V) probs
    P = softmax(Z, axis = 1) # (T, V) probs

    return Z, P

In [34]:
def lossAndSoftmaxBackward(y, p):
    return p - y

In [35]:
dlossBydz = lossAndSoftmaxBackward(y, p)
dlossBydz

array([[ 0.1,  0.2, -0.6,  0.1,  0. ,  0.1],
       [ 0.1,  0.1,  0.2,  0. ,  4. ,  0.1]])